bin/jobserver

```diff
@@ -0,0 +1,453 @@
+from time import sleep
+
+import portage.process
+from portage.util._ctypes import find_library, LoadLibrary
+
+
+__address__ = "127.0.0.1"
+__port__ = 25721
+__max_jobs__ = 1
+__job_delay__ = 0
+__min_free_mem__ = 2097152
+__max_load__ = None
+__parent_pid__ = None
+__server__ = None
+__primary_server__ = None
+
+
+__job_assign_lock__ = threading.Lock()
+__job_counter_lock__ = threading.Semaphore(10240)
+__job_free_slot_lock__ = threading.Semaphore(__max_jobs__)
+__job_running_lock__ = threading.Semaphore(__max_jobs__)
+__last_active__ = datetime.datetime.utcnow()
+__jobs_status__ = ""
+
+__parent_status__ = ""
+__show_status__ = False
+
+__has_ipv6 = None
+
+# from /usr/include/bits/sched.h
+CLONE_NEWNS = 0x00020000
+CLONE_NEWUTS = 0x04000000
+CLONE_NEWIPC = 0x08000000
+CLONE_NEWPID = 0x20000000
+CLONE_NEWNET = 0x40000000
+
+
+class JobServer(BaseRequestHandler):
+    def handle(self):
+        global __job_assign_lock__
+        global __job_counter_lock__
+        global __job_free_slot_lock__
+        global __job_running_lock__
+        global __last_active__
+
+        __job_counter_lock__.acquire()
+        __last_active__ = datetime.datetime.utcnow()
+
+        with __job_assign_lock__:
+            __job_free_slot_lock__.acquire()
+            self._allow_job()
+            __job_running_lock__.acquire()
+            self.request.sendall(b"OK")
+            sleep(__job_delay__)
+
+        while True:
+            try:
+                data = self.request.recv(10)
+            except Exception:
+                break
+            if not data:
+                break
+
+    def finish(self):
+        global __job_counter_lock__
+        global __job_free_slot_lock__
+        global __job_running_lock__
+        global __job_delay__
+
+        __job_running_lock__.release()
+        __job_counter_lock__.release()
+        __job_free_slot_lock__.release()
+
+    def _allow_job(self):
+        global __job_running_lock__
+        global __min_free_mem__
+        global __max_load__
+        global __jobs_status__
+        global __parent_status__
+
+        wait = True
+        while wait:
+            wait = False
+            if __job_running_lock__._value == __max_jobs__:
+                # Allow at least one job to run
+                break
+
+            if __min_free_mem__ is not None:
+                with open("/proc/meminfo") as f:
+                    for line in f.readlines():
+                        if line.startswith("MemAvailable:"):
+                            if int(line.split()[1]) < __min_free_mem__:
+                                __jobs_status__ = "Insufficient Free Memory"
+                                wait = True
+                            break
+
+            if not wait and __max_load__ is not None:
+                if os.getloadavg()[0] > __max_load__:
+                    __jobs_status__ = "System Load Too High"
+                    wait = True
```

Comment:
As it is right now, the `while wait` loop should terminate by default since it's basically a do-while loop. So if `os.getloadavg()[0] <= __max_load__` then `wait` continues to be `False`.

Reply:
Ahh, I see that now. Thanks for the explanation.